ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,314 @@
+    addReplicasToPool(allEligibleReplicas);
+    this.otIterator = new ParanoidDurabilityTrackerIterator();
+  }
+
+  private void addReplicasToPool(List<? extends ReplicaId> replicas) {
+    Collections.shuffle(replicas);
+
+    for (ReplicaId replicaId : replicas) {
+      if (!replicaId.isDown()) {
+          addToBeginningOfPool(replicaId);
+      } else {
+        addToEndOfPool(replicaId); // As a last ditch attempt, we may still try to write to a down replica.
+      }
+    }
+
+    if(!enoughReplicas()) {
+      throw new IllegalArgumentException(generateErrorMessage(partitionId));
+    }
+    initializeTracking();
+  }
+
+  private void initializeTracking() {
+    replicaSuccessCountByDc = new HashMap<String, Integer>();
+    replicaInPoolOrFlightCountByDc = new HashMap<String, Integer>();
+    remoteDatacenters = new ArrayList<String>();
+
+    replicaPoolByDc.forEach((dcName, v) -> {
+      failedCountByDc.put(dcName, 0);
+      replicaSuccessCountByDc.put(dcName, 0);
+      replicaInPoolOrFlightCountByDc.put(dcName, 0);
+      disabledCountByDc.put(dcName, 0);
+      totalReplicaCountByDc.put(dcName, v.size());
+      allDatacenters.add(dcName);
+      if(!dcName.equals(datacenterName)) {
+        remoteDatacenters.add(dcName);
+      }
+    });
+  }
+
+  private String generateErrorMessage(PartitionId partitionId) {
+    StringBuilder errMsg = new StringBuilder("Partition " + partitionId + " has too few replicas to satisfy paranoid durability requirements. ");
+    replicaPoolByDc.forEach((dcName, v) -> {
+      errMsg.append("Datacenter: ").append(dcName).append(" has ").append(v.size()).append(" replicas in LEADER or STANDBY. ");
+    });
+    return errMsg.toString();
+  }
+
+  private boolean enoughReplicas() {
+    for (String currentDc : replicaPoolByDc.keySet()) {
+      if(currentDc.equals(datacenterName)) {
+        if (replicaPoolByDc.get(currentDc).size() < localReplicaSuccessTarget) {
+          logger.error("Not enough replicas in local data center for partition " + partitionId +
+              " to satisfy paranoid durability requirements " +
+              "(wanted " + localReplicaSuccessTarget + ", but found " + replicaPoolByDc.get(currentDc).size() + ")");
+          return false;
+        }
+      } else {
+        if (replicaPoolByDc.get(currentDc).size() < remoteReplicaSuccessTarget) {
+          logger.error("Not enough replicas in remote data center " + currentDc + " for partition " + partitionId
+              + " to satisfy paranoid durability requirements " +
+              "(wanted " + remoteReplicaSuccessTarget + ", but found " + replicaPoolByDc.get(currentDc).size() + ")");
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  public boolean hasSucceeded() {
+    return hasSucceededLocally() && hasSucceededRemotely();
+  }
+
+  private boolean hasSucceededLocally() {
+    return localReplicaSuccessCount >= localReplicaSuccessTarget;
+  }
```

Comment:
we should cconsider dynamic target for local success here, just like the `hasFailed` method.

Reply:
Yeah, I had misunderstood the dynamic target and its relevance, so updated the local success and failure computations.